[Backend Configuration Ia] Pydantic models for communicating backend-independent dataset information

CodyCBakerPhD opened the pull request:
replaces #549 

~250 new lines of source code
~600 lines of tests

Introduces Pydantic models for modelling fields of in-memory neurodata objects that are destined to become Datasets (of h5py or zarr variety) in a way that is agnostic to which backend will ultimately be chosen

Pydantic is used here for many key benefits, one the best of which is automatic JSON schema generation for handling this information to generate forms for the NWB GUIDE

Instantaneous and thorough validation of inputs is also a big plus when there are this many dictionary-like values being passed around - many of the problems that could result from bad inputs to fields like chunk shapes would only be caught at or near the final step of writing, which in this case could be well after specification by the user

In the long-term, using autodoc-pydantic could also allow us to automatically generate formatted docstrings, sphinx API docs, and Entity-Resource graphs (similar to UML) all automatically while still preserving the ability for developers to quickly read and parse the signatures of classes and methods

The design of inputs to the dataset models is also to be friendly and flexible as possible to users of the API; for those using YAML feature or simple string specification, we automatically fetch and instantiate any complicated `hdf5plugin`/`numcodec` objects from the string inputs; and for those who already know how to instantiate such items, they may pass them directly

h-mayorquin commented on src/neuroconv/tools/testing/_mock/_mock_dataset_models.py:
```diff
@@ -0,0 +1,57 @@
+)
+
+
+def mock_DatasetInfo() -> DatasetInfo:
```
Why make this a function and not just a variable?

CodyCBakerPhD replied:
Following the style of the PyNWB mocks: https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/src/pynwb/testing/mock/base.py

It definitely would not be a variable, I would appreciate a separate instance in memory each time

Probably not as relevant for the immutable one but very important for the mutable one

h-mayorquin commented on src/neuroconv/tools/nwb_helpers/_models/_base_dataset_models.py:
```diff
@@ -0,0 +1,115 @@
+
+        return values
+
+    def get_data_io_keyword_arguments(self):
```
Or maybe is not even needed.

h-mayorquin replied:
Oh, I see that in another PR you want to have a uniform method to call this. I need to take a closer look.

CodyCBakerPhD replied:
I suppose it could be; it is indeed required to be used in the `configure_backend` method: https://github.com/catalystneuro/neuroconv/pull/571/files#diff-91679913ea0f31861b6d5e7275fa4a7e03b256d6cd91aa8d82ca505c604d9112R226

Went ahead and made it abstract (that is, required to be defined for all child classes)

h-mayorquin commented on src/neuroconv/tools/nwb_helpers/_models/_hdf5_dataset_models.py:
```diff
@@ -0,0 +1,76 @@
+        default=None, description="The optional parameters to use for the specified compression method."
+    )
+
+    def get_data_io_keyword_arguments(self) -> Dict[str, Any]:
```
```suggestion
    def get_data_io_keyword_arguments(self) -> Dict[str, Any]:
    """
    You can use like this
    io = H5DataIO(**get_data_io_keyword_arguments())
    """
```
Can you add some clarification to the docstring like this. Also for Zarr.

This is one of the few instances that I am tempted to use `_kwargs` instead of `keyword_arguments`.

h-mayorquin replied:
I see that is meant to be used with the `set_data_io` container method. I need to think more about this. I also read:

data_io_kwargs = dataset_configuration.get_data_io_kwargs()` in another PR. But could not locate that function.

CodyCBakerPhD replied:
https://github.com/catalystneuro/neuroconv/pull/571/files#diff-91679913ea0f31861b6d5e7275fa4a7e03b256d6cd91aa8d82ca505c604d9112R226

h-mayorquin replied:
Thanks, just to clarify. Do you change the name in another PR further down the chain from `get_data_io_keyword_arguments` to `get_data_io_kwargs()` ?

CodyCBakerPhD replied:
Looks like it is still the more verbose version: https://github.com/catalystneuro/neuroconv/blob/integrate_with_interfaces_and_converter/src/neuroconv/tools/nwb_helpers/_models/_hdf5_models.py#L53

CodyCBakerPhD replied:
I prefer the more verbose version to help external readers understand it faster

h-mayorquin commented on src/neuroconv/tools/nwb_helpers/_models/_base_dataset_models.py:
```diff
@@ -0,0 +1,115 @@
+        return hash((type(self),) + tuple(self.__dict__.values()))
+
+    def __str__(self) -> str:
+        """Not overriding __repr__ as this is intended to render only when wrapped in print()."""
```
Can you elaborate on why we need to modify __str__ but you need to keep __repr__?

Like, obviously you want to keep __repr__ intact for some reason but the reason here is not clear.

CodyCBakerPhD replied:
Added enhanced description

Reason being two-fold; a standard `repr` is intended to be slightly more machine readable / a more basic representation of the true object state. But then also because an iterable of these objects, such as a `List[DataSetInfo]`, would print out the nested representations, which only look good when using the basic `repr` (that is, this fancy string print-out does not look good when nested in another container).

h-mayorquin replied:
Thanks. Most of this was resolved in an in-person talk. So, this still does not play as intended in a notebook because notebooks use __repr__ when outputing a value instead of print.

I don't think that solving this to have good representation in all possible environments is a good use of our  time. So let's leave it as it is, right now you are the only playing with this and I think it makes sense that you chooose the representation that you find more useful. We can come back to this in the future if needed.

h-mayorquin commented on src/neuroconv/tools/nwb_helpers/_models/_zarr_dataset_models.py:
```diff
@@ -0,0 +1,122 @@
+
+from ._base_dataset_models import DatasetConfiguration
+
+_available_zarr_filters = (
```
Your milleage might vary but I would define sets with names like

```
loosy_codes = {"astype", "bitround", "quantize"}   # 
```
And then substract them.

I find the nested structure with comments and initilaization in the fly hard to read. Variables are very cheap and with good naming work as documentation.

CodyCBakerPhD replied:
Enhanced the readability here

h-mayorquin replied:
Thanks.

h-mayorquin reviewed (commented):
Did the second readin of this. Now I left more comments. I will take another look after as this is not a priority but I want to keep the discussion going so we can have time once the upstream changes are approved.

CodyCBakerPhD commented on src/neuroconv/tools/nwb_helpers/_models/_base_dataset_models.py:
```diff
@@ -0,0 +1,115 @@
+        arbitrary_types_allowed = True
+
+    dataset_info: DatasetInfo = Field(description="The immutable information about this dataset.")
+    chunk_shape: Tuple[int, ...] = Field(
```
We must pin to https://docs.pydantic.dev/1.10/ for now

But yes, when DANDI goes to 2.0 we can easily adjust to that

Would you prefer I leave a comment to that effect?

h-mayorquin replied:
I see. I was just not aware this was a pydantic 2.0 thing. A comment would be a useful reminder though.

CodyCBakerPhD replied:
I added a comment

CodyCBakerPhD commented on src/neuroconv/tools/nwb_helpers/_models/_hdf5_dataset_models.py:
```diff
@@ -0,0 +1,76 @@
+    _available_hdf5_filters = _available_hdf5_filters | set(
+        (str(hdf5plugin_filter).rstrip("'>").split(".")[-1] for hdf5plugin_filter in hdf5plugin.get_filters())
+    )  # Manual string parsing because of slight mismatches between .filter_name and actual import class
+AVAILABLE_HDF5_COMPRESSION_METHODS = tuple(_available_hdf5_filters)
```
It's a convention for global variables, which environment variables are one type of

CodyCBakerPhD replied:
Parts of the HDF5 plugin mechanism are even determined via other environment variables

h-mayorquin replied:
All right. Thanks, it was just more curiosity on my side. Thanks for this.

CodyCBakerPhD commented on src/neuroconv/tools/nwb_helpers/_models/_hdf5_dataset_models.py:
```diff
@@ -0,0 +1,76 @@
+    import hdf5plugin
+
+    _available_hdf5_filters = _available_hdf5_filters | set(
+        (str(hdf5plugin_filter).rstrip("'>").split(".")[-1] for hdf5plugin_filter in hdf5plugin.get_filters())
```
Depends on the users system. You could define or dynamically import any arbitrary number of them, but they do all get registered along this pathway, which is how the plugin library detects them

CodyCBakerPhD replied:
Some are also only available on certain platforms/versions/architectures

h-mayorquin replied:
I mean the mapping between  what is available  in `hdf5plugin.get_filters()` to the variable  `_available_hdf5_filters`. I am proposing just having a translation dict instead of the parsing you have here. It is easier to change and update if there are changes in what you get in `hdf5plugin.get_filters()` or if there is an error. Though, I think this work if the number of elements is <15 and after that it becomes too much.

CodyCBakerPhD replied:
This is a design of how it ought to look in the GUIDE as a dropdown selection

My results from `_available_hdf5_filter`:

```python
{'BZip2',
 'Bitshuffle',
 'Blosc',
 'Blosc2',
 'FciDecomp',
 'LZ4',
 'SZ',
 'SZ3',
 'Zfp',
 'Zstd',
 'gzip',
 'lzf'}
```

Without this nicer parsing it looks like this

```
["<class 'hdf5plugin._filters.Bitshuffle'>",
 "<class 'hdf5plugin._filters.Blosc'>",
 "<class 'hdf5plugin._filters.Blosc2'>",
 "<class 'hdf5plugin._filters.BZip2'>",
 "<class 'hdf5plugin._filters.FciDecomp'>",
 "<class 'hdf5plugin._filters.LZ4'>",
 "<class 'hdf5plugin._filters.SZ'>",
 "<class 'hdf5plugin._filters.SZ3'>",
 "<class 'hdf5plugin._filters.Zfp'>",
 "<class 'hdf5plugin._filters.Zstd'>"]
```

Oh also, the base filters do NOT have that same class form since they are built-in to `h5py`, you just refer to them with the 'parsed' string form such as `"gzip"`

CodyCBakerPhD replied:
Or are you simply suggesting to turn the entier thing (the zarr one too) into a dict where the keys are the parsed versions and the values are the actual classes?

CodyCBakerPhD replied:
Sure, will do

CodyCBakerPhD commented on src/neuroconv/tools/nwb_helpers/_models/_base_dataset_models.py:
```diff
@@ -0,0 +1,131 @@
+        arbitrary_types_allowed = True
+
+    dataset_info: DatasetInfo = Field(description="The immutable information about this dataset.")
+    chunk_shape: Tuple[int, ...] = Field(
```
```suggestion
    chunk_shape: Tuple[int, ...] = Field(  # When using Pydantic v2.0, specify PositiveInt
```

h-mayorquin reviewed (commented):
This looks good ot me. It is very nice that we have this unification for the two backends with pydantic and we can validate the parameters in a general way.

I left more comments and I make two big points here that are more general:

Do we really need two classes for `DatasetInfo` and `DatasetConfiguration`. While conceptually they are different things (as dataset stands on its own)  you seem to be using them together:

https://github.com/catalystneuro/neuroconv/blob/0058e35becdc2638643ad25e354354510e0fbd86/src/neuroconv/tools/nwb_helpers/_dataset_configuration.py#L106

We would have less maintenance effort both in testing and lines on code but also in the number of concepts that we have in the library if we keep kept `DatasetConfiguration` and handle everything from there. You already have the options for `compression_options` and `compression_method` to be None which seem to me represent the pure dataset quite well.

 The second point is that you introduced some code here for `get_data_io_keyword_arguments` that we are not testing yet. I think it would have been better to delay the introduction of this until a further PR so I can see its use and tests at the same time.

Besides those two points that I think are not critical this looks ready to me.

CodyCBakerPhD commented on src/neuroconv/tools/nwb_helpers/_models/_base_dataset_models.py:
```diff
@@ -0,0 +1,131 @@
+    object_id: str = Field(description="The UUID of the neurodata object containing the dataset.")
+    location: str = Field(description="The relative location of the this dataset within the in-memory NWBFile.")
+    full_shape: Tuple[int, ...] = Field(description="The maximum shape of the entire dataset.")
+    dtype: np.dtype = Field(  # TODO: When using Pydantic v2, replace np.dtype with InstanceOf[np.dtype]
```
This is intentional - this is really a machine generated container intended to be returned by a function, I would never expect or want a user to hand type an instance of it (especially the object ID), or if they did for whatever reason I would avoid duck typing because that is WAY harder to validate in JSON because I'd have to compile a string listing of all allowed inputs to Dtype (if you know an easy way to get that I'm all ears)

h-mayorquin replied:
Got you. 
We don't leave that level of generality so why pay for it? If our use case is solved by only using np.dtype then we should just use that.

CodyCBakerPhD commented on tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_zarr_dataset_configuration_model.py:
```diff
@@ -0,0 +1,280 @@
+    assert "gzip" in AVAILABLE_ZARR_COMPRESSION_METHODS
+
+
+def test_mutation_validation():
```
That was my thining too - not exhaustive testing since I assume Pydantic handles that - but for my own sanity I want to make sure it is enabled properly by us

h-mayorquin replied:
Yes. I this line of reasoning makes sense to me.

CodyCBakerPhD commented on tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_zarr_dataset_configuration_model.py:
```diff
@@ -0,0 +1,280 @@
+
+    expected_error = (
+        "Length mismatch between `filter_methods` (2 methods specified) and `filter_options` (1 options found)! "
+        "These two must match one-to-one. (type=value_error)"
```
Sure

CodyCBakerPhD commented on src/neuroconv/tools/testing/_mock/_mock_dataset_models.py:
```diff
@@ -0,0 +1,57 @@
+def mock_ZarrDatasetConfiguration(
+    compression_method: Union[Literal[AVAILABLE_ZARR_COMPRESSION_METHODS], numcodecs.abc.Codec, None] = "gzip",
+    compression_options: Union[Dict[str, Any]] = None,
+    filter_methods: Iterable[Union[Literal[AVAILABLE_ZARR_COMPRESSION_METHODS], numcodecs.abc.Codec, None]] = None,
```
The main point being because the `AVAILABLE_BACKEND_COMPRESSION_METHODS` are not written explicitly but instead inferred programatically?

h-mayorquin replied:
Looking at the source code is hard for me to know what kind of input this function accepts. The default value being "gzip" helps a lot though.

I was also thinking that the help that is displayed is not useful but when you actually use the help method it becomes useful. Look at what vscode shows top vs what the help method shows.

![image](https://github.com/catalystneuro/neuroconv/assets/6429509/2042bc64-22ca-4ff4-9526-36fb77cb240b)

CodyCBakerPhD replied:
Huh, the `Any` is an interesting interpretation there

Any way you can think of to improve the reading of the source code?

h-mayorquin replied:
Mmm no ideas, for the datset models classess I think the field annotation is doing a good job. Probably not super relevant for the mock examples.

CodyCBakerPhD commented:
> Following the standard of that makes sense. Would it be a good idea to generate a separate instance of object_id each time?

I actually kind of prefer having it be static, such that if you ever see that object ID in practice, you know there is a near certainty that the object was generated by the mock instead of an actual neurodata object

CodyCBakerPhD commented on src/neuroconv/tools/nwb_helpers/_models/_zarr_dataset_models.py:
```diff
@@ -0,0 +1,124 @@
+        ),
+    )
+    # TODO: actually provide better schematic rendering of options. Only support defaults in GUIDE for now.
+    # Looks like they'll have to be hand-typed however... Can try parsing the numpy docstrings - no annotation typing.
```
Nah this is still a 'to-do' eventual feature perhaps; this refers to the input arguments that any individual codec would receive as 'valid'

such as 'level' of `Gzip`

But they don't use annotation typing in the signatures so that makes it harder

CodyCBakerPhD replied:
Maybe hand code table for set of core supported ones but otherwise fallback to open string

h-mayorquin commented on src/neuroconv/tools/nwb_helpers/_models/_base_dataset_models.py:
```diff
@@ -0,0 +1,131 @@
+        if len(chunk_shape) != len(buffer_shape):
+            raise ValueError(f"{len(chunk_shape)=} does not match {len(buffer_shape)=}!")
+        if len(buffer_shape) != len(full_shape):
+            raise ValueError(f"{len(buffer_shape)=} does not match {len(full_shape)=}!")
```
```suggestion
            raise ValueError(f"{len(buffer_shape)=} does not match {len(full_shape)=}!  for dataset located in {self.dataset_info.location} ")
```

When running a transformation with multiple datasets is not going to be obvious which dataset is causing the problems. We should add the information of the dataset to the user so they know where exactly to fix this. We should do the same for the other assertions here.

CodyCBakerPhD replied:
Done

CodyCBakerPhD commented:
@h-mayorquin I made a few final changes to the string printout and added automatic parsing of the dataset name

Also added a bunch more tests and parameterized a lot of the previous ones to save space

h-mayorquin commented on tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_dataset_configuration_model.py:
I am not sure we need to test these ones as they are basically the promise of the ABC.

CodyCBakerPhD replied:
I still like to for my own sanity; _i.e._, to make sure _we_ used `ABC` correctly

h-mayorquin commented on tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_common_dataset_configuration_model.py:
This is a fine use of parameterize I feel.

h-mayorquin commented on tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_zarr_dataset_configuration_model.py:
```diff
@@ -0,0 +1,212 @@
+    assert "gzip" in AVAILABLE_ZARR_COMPRESSION_METHODS
+
+
+def test_get_data_io_keyword_arguments():
```
Note that in a future PR you change your name to `get_data_io_kwargs` which I think is a better name as referring to keyword arguments is a common python idiom and we use this in other places in the library. Let's change it now from the base so its propagated upwards.

CodyCBakerPhD replied:
Fine, for the sake of getting this through I'll just change to `kwargs`

h-mayorquin commented on src/neuroconv/tools/nwb_helpers/_models/_base_dataset_models.py:
```diff
@@ -0,0 +1,180 @@
+    @abstractmethod
+    def get_data_io_keyword_arguments(self):
+        """
+        Fetch the properly structured dictionary of input arguments to be passed directly into a H5DataIO or ZarrDataIO.
```
```suggestion
        Fetch the properly structured dictionary of input arguments to be passed directly into a H5DataIO or ZarrDataIO. That is, you can wrap data using `wrapped_data=H5DataIO(**get_data_io_kwargs` or `wrapped_data=ZarrDataIO(**get_data_io_kwargs)`.
```

Would it be possible for this to aslo return the data so the above docstring is correct? We use this pattern in the tools to create all the keyword that are passed to create an object. I think we should honor this pattern here as well. 

Another option would be to have the kwargs for  `set_data_io` so we could do:

```python
        nwbfile_objects[object_id].set_data_io( **dataset_configuration.get_data_io_kwargs())
```
Instead of what we have now which is:
```python
        nwbfile_objects[object_id].set_data_io(dataset_name=dataset_name, data_io_class=data_io_class, **data_io_kwargs)
```

This is probably our main interest here as well.

h-mayorquin replied:
Answering to why the suggested docstring is incorrect:

* `H5DataIO` takes the data as well which is not returned by the current function. That is,  `wrapped_data=H5DataIO(**get_data_io_kwargs)` will not work.

I am trying to keep consistency with the pattern that we use in the tools (spikeinterface and roiextractors tools in neuroconv I mean):

Examples:
https://github.com/catalystneuro/neuroconv/blob/9efdefa5963ecca4ea6cd8ff392ec0b6be903702/src/neuroconv/tools/spikeinterface/spikeinterface.py#L679-L680 

https://github.com/catalystneuro/neuroconv/blob/9efdefa5963ecca4ea6cd8ff392ec0b6be903702/src/neuroconv/tools/roiextractors/roiextractors.py#L388-L393

But I actually think that trying to avoid dynamic passing is a worthwhile goal. I am willing to consider other options. I have some ideas but maybe we can discuss that in another PR down the chain?

CodyCBakerPhD replied:
Ah, but these are not intended to be the `kwargs` that get passed into the function `.set_data_io`, but instead they fill the role of the third argument of the `.set_data_io` method

Note that if `wrapped_data=H5DataIO(**get_data_io_kwargs)` works under your model, that might only be because the `data_io_class` keyword argument would be ignored/skipped because it's not in the signature

Hence I prefer mine for readability as well; 'set my data IO on the dataset with `dataset_name` using the `data_io_class` of my choosing, which takes and uses additional keyword arguments structured by `.get_data_io_kwargs()`'

h-mayorquin commented on src/neuroconv/tools/nwb_helpers/_models/_base_dataset_models.py:
```diff
@@ -0,0 +1,180 @@
+        return values
+
+    @abstractmethod
+    def get_data_io_keyword_arguments(self):
```
```suggestion
    def get_data_io_kwargs(self):
```
I think we should change this now. See the comments in the tests for the reasons.

h-mayorquin commented on src/neuroconv/tools/nwb_helpers/_models/_base_dataset_models.py:
```diff
@@ -0,0 +1,180 @@
+from pydantic import BaseModel, Field, root_validator, validator
+
+
+class DatasetInfo(BaseModel):
```
Can you add a small docstring here clarifying how is this different from `DatasetConfiguration(BaseModel,`. If I remember well our in-person discussion you want this object to correspond to the inmutable data so let's keep the docstring as a reminder on why these concepts were separated as a clarification of intent for future developers (versions of us most likely).

CodyCBakerPhD replied:
Done

h-mayorquin replied:
Thanks.

h-mayorquin reviewed (changes requested):
OK, I checked this again with fresh eyes. Thanks for bearing with me @CodyCBakerPhD 

I have a final set of requests about the `get_data_io_kwargs` function and this should be good to go.

I have one final question that is not blocking but I would like to hear your opinion on it. 

I said before in this PR that I woud have liked to have a single class instead of two with `DataSetInfo` and `DataSetConfiguration`. My argument for this was to avoid the profileration of classes and to keep tighly coupled code together. Now, If I remember well, what you said was that you wanted to keep the non-mutable attributes in a separate class. Now, here is my question.

The latest version of pydantic has the concept of inmutability at the field level:
https://docs.pydantic.dev/latest/concepts/fields/#immutability

Which would allow to indicate that the attributes  `object_id`, `location` and `dataset_name` are non mutable.

If that feature were available, would you be OK with the unification? 

Now, because the constraints from Dandi this is not available. there is something similar with `allow_mutation=False` at the field level but I think this was introduced later than our pinned version:

https://github.com/pydantic/pydantic/issues/2195

As this is not posssible with our dependencies as they are I just want to get a sense of how you feel so we can keep this in the back of our minds.

CodyCBakerPhD commented:
> > If that feature were available, would you be OK with the unification?
>
> Yeah, that would be cool

Although, to be aware, at that point you would loose the hashability feature, which relies on the entire object being immutable. Earlier designs of the later PRs utilized this approach until the `location` parsing became so reliable that it wasn't really needed any more... might still be nice to have? The v2 jump isn't happening any time soon though so maybe we'll cross that bridge when we get there
